src/Build/BackEnd/Components/Communications/NodeProviderOutOfProcBase.cs

```diff
@@ -289,6 +289,15 @@ protected NodeContext GetNode(string msbuildLocation, string commandLineArgs, in
                     CommunicationsUtilities.Trace("Successfully connected to created node {0} which is PID {1}", nodeId, msbuildProcess.Id);
                     return new NodeContext(nodeId, msbuildProcess, nodeStream, factory, terminateNode);
                 }
+
+                if (msbuildProcess.HasExited)
+                {
+                    CommunicationsUtilities.Trace($"Could not connect to node with PID {msbuildProcess.Id}; it has exited. This can indicate a crash at startup");
+                }
+                else
+                {
+                    CommunicationsUtilities.Trace($"Could not connect to node with PID {msbuildProcess.Id}; it is still running. This can occur when two multiprocess builds run in parallel and the other one 'stole' this node");
```

Comment:
It would be nice if here, or when failing to connect, we could log whether the pipe exists but is busy vs pipe does not exist. Unfortunately NamedPipeClientStream.Connect throws TimeoutException for both, so it's not possible to distinguish.

https://github.com/danmoseley/runtime/blob/2db2fa13cf7012b8c16a51cd982a4f536d5c444e/src/libraries/System.IO.Pipes/src/System/IO/Pipes/NamedPipeClientStream.Windows.cs#L71-L75
https://github.com/danmoseley/runtime/blob/2db2fa13cf7012b8c16a51cd982a4f536d5c444e/src/libraries/System.IO.Pipes/src/System/IO/Pipes/NamedPipeClientStream.Unix.cs#L43-L46
https://github.com/danmoseley/runtime/blob/2db2fa13cf7012b8c16a51cd982a4f536d5c444e/src/libraries/System.IO.Pipes/src/System/IO/Pipes/NamedPipeClientStream.cs#L171-L172

It might be nice to make that exception message include that info. @JeremyKuhne thoughts, am I missing any way in managed code to determine whether a pipe exists but is busy?

Reply:
> am I missing any way in managed code to determine whether a pipe exists but is busy?

Not that I'm aware of. Setting the `ErrorCode` to something more useful might be an option.